### ISSUE TITLE
Estimation of memory usage (SparXCC_base)

### ISSUE THREAD

**Issue by SantosRAC** [IbTJensen/CompoCor#1]
Dear @IbTJensen ,
I hope you are doing well.

I was wondering if you have any tip on how to estimate the amount of RAM a job would use with `SparXCC_base`. I am trying to run it with 6 cores and 100 or 200Gb of RAM, but it looks like I am still running out of memory.

My two matrices are 176 samples (3000 OTUs and 39k genes).

All the best

**Comment by IbTJensen** [comment]
Dear @SantosRAC ,
Thank you for your interest in my package.

Firstly, I would highly recommend some kind of prevalence based filtering. There are several reasons for:
1. SparXCC does not scale up well in terms of running time and memory usage (as you have already noticed).
2. SparXCC does not provide useful estimates of correlations involving variables with low read counts.
3. The dynamic permutation threshold will be more conservative the more variables are included.
By removing variables with low read counts your problem becomes computationally feasible, you mainly lose estimates that would have been unreliable anyway, and you will most likely pick up more of the reliable signals in the data.

I have not done an in-depth analysis of the memory usage, but the running time is dominated by the number of variables of the bigger dataset (the number of genes in your case), so this is most likely the case for memory usage as well. I believe the memory usage to be O(q^2), where q is the number of genes. This means that a dataset with twice as many genes will require 4 times more memory. To estimate the required memory for the full data, I would run SparXCC on a subset of the data (for example all OTUs, all data points and 5k genes), note the memory usage, and then scale up accordingly (in this case it would require roughly 8^2 = 64 times more memory for all 39k genes).

I hope this helps. Feel free to ask any followup question as needed, and good luck with your analysis!

Best,
Ib

**Comment by SantosRAC** [comment]
Dear @IbTJensen ,

Thank you very much for all your help and advice. They were really helpful.

For FIY: last week I tried to run SparXCC_base with the 39K, and I was not able to finish a job even with 990Gb available in a high memory queue of our cluster.

Today, I filtered out low expressed genes and low variance (not absolutely sure this helps) and I ran it with a 7K genes x 3k OTUs matrix. It finished very quickly, apparently it was successful.

I have another questions though:

This is how I ran (requesting 8 cores and 500Gb, but with max memory usage = 36.18GiB):

```R
SparXCC_base(wallace_otu_counts_day_filtered_transposed,
                        kremling_expression_counts_day_filtered_transposed,
                        pseudo_count = 1,
                        var_min = 1e-05,
                        Find_m = TRUE,
                        B_m = 100,
			cores = 8)
```

The final output has genes (columns) and OTUs (rows), and rows are partially filled by `FALSE`. I understand it mean FALSE correspond to cases where `m` (permutation threshold) did not pass the threshold. Perhaps it could be nice to print or generate a second object with statistics like the `m` itself.

Regarding the correlation values, I'm just picking the first OTU as example. Values range from `-0.394849369220677` to `0.508644507056246`. How do these values compare with linear Pearson correlations? Do you have any recommendation besides an arbitrary threshold (values above/under certain values are would be highly correlated)?

The last question: besides dealing with compositionality and sparsity, I was wondering if outlier values have also been considered in your work.

Thank you very much one more time for this nice tool!

**Comment by SantosRAC** [comment]
@IbTJensen I'm sorry. I believe now I understand:

1. I was trying to undertand why part of my genes started with "cor" and the othe part started with "correlted". Now I see that half of the columns are correlation values (with OTUs in rows) and the other half is TRUE/FALSE (whether that pair passed the `m threshold`);
2. In my case, using the command above (with gene and OTU counts that passed some pre-filtering steps), I have just a very few "TRUE" cells compared to total.

Also, forget about my example of correlation values above (`-0.394849369220677` to `0.508644507056246`). In fact, I have almost nothing considered truly correlated in my datasets, if my current understanding is correct.

I will write some code to parse this result matrix, but if you already have some code available, this could be of interest to other users.

All the best,
Renato

**Comment by SantosRAC** [comment]
Only one thing that is still weird to me: I could not figure out why the number of columns of "correlated" do not match the number of column of "cor". Is there any way I can share big matrices with you? (unpublished results)

EDIT: Got it. Found the "m" value (column between "cor" and "correlated"). I'm sorry about the amount of messages :-) I will go ahead and close the issue after a few days.

**Comment by IbTJensen** [comment]
Dear @SantosRAC 

It looks like you worked out your problem by yourself, but I'll answer some of your questions regardless.

The output should be a list (say L) where L$cor contains the estimated cross-correlation matrix between the OTUs and the genes, L$m contains the permutation threshold and L$correlated indicates if a given cross-correlation exceeds m. It does not seem like this is reflected in the documentation at the moment.

Would the output be more intuitive if "correlated" had a more desciptive name, say "cor_above_m"?

Regarding comparisons to other methods, the correlations estimated by SparXCC should be interpreted as approximations of the linear Pearson correlations between the absolute abundances and absolute expression levels. How accurate this approximations is, depends on the level of sparsity of the true correlations matrix.

Regarding the small number of truly correlated pairs, this may in part be due to the number of variables. The threshold m represents the maximum expected correlation estimated by SparXCC, if all pairs had been uncorrelated. Remember that you still have 21 million pairs, so estimates could still get fairly large due to chance. You may consider more stringent criteria for filtering your OTUs/genes. Whether or not this will make a difference will depend on your data.

Best regards,
Ib

**Comment by SantosRAC** [comment]
Hello @IbTJensen ,

Thanks for your detailed explanation. I'm really using your software and I hope I can present some nice results pretty soon!

Regarding your explanation about the list (L), it was actually my mistake again. Instead of exporting only L$cor or L$correlated, I realized I exported the whole L. Anyway, since I already developed some code to parse this list (a big matrix with cor + m + correlated boolean), I will keep this for my work. I believe the way your code is implemented now is fine!

PS: yes, "cor_above_m" could help, but I would not worry about it because "correlated" also implies there was some kind of criteria to keep them, in your case "passing the m threshold" (I think this issue will be helpful to other folks).
